contrib/win32/win32compat/win32_usertoken_utils.c

```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
 	return s_lsa_auth_pkg;
 }
 
+int lookup_principal_name(const wchar_t * sam_account_name, wchar_t * user_principal_name)
```

Comment:
recommend making this a void func and adding a comment that it provides best effort upn resolution